data-prepper-core/src/main/java/org/opensearch/dataprepper/peerforwarder/PeerForwarder.java

```diff
@@ -0,0 +1,175 @@
+                try {
+                    serializedJsonString = serializeRecordsToJsonString(entry.getValue());
+                } catch (JsonProcessingException e) {
+                    throw new RuntimeException(e);
+                }
+
+                final HttpRequest httpRequest = createHttpRequest(serializedJsonString);
+                final CompletableFuture<AggregatedHttpResponse> aggregatedHttpResponseCompletableFuture = forwardRecords(client, httpRequest);
+
+                try {
+                    final AggregatedHttpResponse response = aggregatedHttpResponseCompletableFuture.get();
+                    if (response == null || response.status() != HttpStatus.OK) {
+                        recordsToProcessLocally.addAll(entry.getValue());
+                    }
+                } catch (InterruptedException | ExecutionException e) {
+                    LOG.error("Problem with asynchronous peer forwarding", e);
+                }
+            }
+        }
+        return recordsToProcessLocally;
+    }
+
+    private Map<String, List<Record<Event>>> groupRecordsBasedOnIdentificationKeys(Collection<Record<Event>> records, Set<String> identificationKeys) {
+        final Map<String, List<Record<Event>>> groupedRecords = new HashMap<>();
+
+        // group records based on IP address calculated by HashRing
+        for (final Record<Event> record : records) {
+            final Event event = record.getData();
+
+            List<String> identificationKeyValues = new LinkedList<>();
+            for (final String identificationKey : identificationKeys) {
+                identificationKeyValues.add(event.get(identificationKey, Object.class).toString());
+            }
+
+            final String dataPrepperIp = hashRing.getServerIp(identificationKeyValues).orElse(StaticPeerListProvider.LOCAL_ENDPOINT);
+            groupedRecords.computeIfAbsent(dataPrepperIp, x -> new ArrayList<>()).add(record);
+        }
+        return groupedRecords;
+    }
+
+    private WebClient getClient(final String address) {
+        return isAddressDefinedLocally(address) ? LOCAL_CLIENT : peerClientPool.getClient(address);
+    }
+
+    private boolean isAddressDefinedLocally(final String address) {
+        final InetAddress inetAddress;
+        try {
+            inetAddress = InetAddress.getByName(address);
+        } catch (UnknownHostException e) {
+            return false;
+        }
+        if (inetAddress.isAnyLocalAddress() || inetAddress.isLoopbackAddress()) {
+            return true;
+        } else {
+            try {
+                return NetworkInterface.getByInetAddress(inetAddress) != null;
+            } catch (SocketException e) {
+                return false;
+            }
+        }
+    }
+
+    private boolean isLocalClient(final WebClient client) {
+        return client == LOCAL_CLIENT;
+    }
+
+    private String serializeRecordsToJsonString(final List<Record<Event>> records) throws JsonProcessingException {
```

Comment:
Moving serialization into its own class would be a good approach here. You can have a single class that is responsible for both serialization and deserialization. It will be easier to write unit tests against both `PeerForwarder` and the serializer class.